aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -103,18 +108,83 @@ compute_result_type(at::ArrayRef<OperandInfo> operands,
 
 std::tuple<Device, ScalarType> TensorIterator::compute_common_type() {
   // See [Result type computation] in TensorIterator.h
+
   auto result_type =
       compute_result_type(operands_,
-        [](const Tensor& t) { return t.dim() > 0; },
-        [](const Tensor& t) { return !t.unsafeGetTensorImpl()->is_wrapped_number(); },
-        [](const Tensor& t) { return true; });
+        [](const OperandInfo& op) { return op.tensor.dim() > 0; },
+        [](const OperandInfo& op) { return !op.tensor.unsafeGetTensorImpl()->is_wrapped_number(); },
+        [](const OperandInfo& op) { return true; });
+
+  if (ScalarType::Bool == std::get<1>(result_type)) {
+    auto alternate = compute_result_type(operands_,
+        [](const OperandInfo& op) {
+          return op.tensor.dim() == 0;
+        }
+    );
+    if (std::get<1>(alternate) != ScalarType::Undefined) {
+      // preserve device from original result
+      return std::make_tuple(std::get<0>(result_type), std::get<1>(alternate));
+    }
+  }
+
+  // if non-zero-dim tensor result is an integral type and there's a zero-dim
+  // floating point operand, we'll promote the floating point type.
+  if (isIntegralType(std::get<1>(result_type), false)) {
+    auto alternate = compute_result_type(operands_,
+        [](const OperandInfo& op) {
+          return isFloatingType(op.tensor.scalar_type()) && op.tensor.dim() == 0;
+        }
+    );
+    if (std::get<1>(alternate) != ScalarType::Undefined) {
+      // preserve device from original result
+      return std::make_tuple(std::get<0>(result_type), std::get<1>(alternate));
+    }
+  }
 
   TORCH_INTERNAL_ASSERT(std::get<1>(result_type) != ScalarType::Undefined);
   return result_type;
 }
 
+static void validate_dtype(OperandInfo& op, ScalarType common_dtype, int ninputs) {
+  if (op.tensor.defined()) {
+    // For binary_ops, we follow casting rules. For unary/nullary types
+    // we require the type to match.
+    if (op.is_output) {
+      if ((ninputs < 2 && op.dtype != op.tensor.scalar_type()) ||
+          !canCast(common_dtype, op.tensor.scalar_type()))
+      {
+        AT_ERROR("result type ", common_dtype,
```

Comment:
can we have different error messages for these 2 cases?  In the first case, it's not that it can't be cast, it that it won't case.

Reply:
moved the code around to simplify and use the below message. the ninputs<2 case is the same regardless of if it's an output tensor.